filters/apimonitoring/filter.go

```diff
@@ -0,0 +1,158 @@
+package apimonitoring
+
+import (
+	"github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	metricsSeparator = "."
+	metricCountAll   = "http_count"
+	metricCount500s  = "http5xx_count"
+	metricCount400s  = "http4xx_count"
+	metricCount300s  = "http3xx_count"
+	metricCount200s  = "http2xx_count"
+	metricLatency    = "latency"
+)
+
+const (
+	stateBagKeyPrefix = "filter.apimonitoring."
+	stateBagKeyState  = stateBagKeyPrefix + "state"
+)
+
+var (
+	unknownPath = &pathInfo{
+		ApplicationId: unknownElementPlaceholder,
+		ApiId:         unknownElementPlaceholder,
+		PathTemplate:  unknownElementPlaceholder,
+	}
+)
+
+type apiMonitoringFilter struct {
+	paths []*pathInfo
+}
+
+type pathInfo struct {
+	ApplicationId string
+	ApiId         string
+	PathTemplate  string
+	Matcher       *regexp.Regexp
+}
+
+// apiMonitoringFilterContext contains information about the metrics tracking
+// for one HTTP exchange (one routing). It serves to pass information from
+// the `Request` call to the `Response` call (stored in the context's `StateBag`).
+type apiMonitoringFilterContext struct {
+	// DimensionPrefix is the prefix to all metrics tracked during this exchange (generated only once)
+	DimensionsPrefix string
+	// Begin is the earliest point in time where the request is observed
+	Begin time.Time
+	// OriginalRequestSize is the initial requests' size, before it is modified by other filters.
+	OriginalRequestSize int64
+}
+
+func (f *apiMonitoringFilter) String() string {
+	return toJsonStringOrError(mapApiMonitoringFilter(f))
+}
+
+func (f *apiMonitoringFilter) Request(c filters.FilterContext) {
+	log := log.WithField("op", "request")
+
+	// Identify the dimensions "prefix" of the metrics.
+	dimensionsPrefix := f.getDimensionPrefix(c, log)
+
+	// Gathering information from the initial request for further metrics calculation
+	begin := time.Now()
+	originalRequestSize := c.Request().ContentLength
+
+	// Store that information in the FilterContext's state.
+	mfc := &apiMonitoringFilterContext{
+		DimensionsPrefix:    dimensionsPrefix,
+		Begin:               begin,
+		OriginalRequestSize: originalRequestSize,
+	}
+	c.StateBag()[stateBagKeyState] = mfc
+}
+
+func (f *apiMonitoringFilter) Response(c filters.FilterContext) {
+	log := log.WithField("op", "response")
+	log.Debugf("RESPONSE CONTEXT: %s", c)
```

Comment:
I think, although for debug mode, it is a huge overhead to log four lines with lots of unnecessary information for one only one api usage monitoring event. Please restrict this to one log line with only the essential information, i.e. the `req.URL`, the generated `prefix`, the response status code, and may be the latency.

Reply:
That was for detailed low-level debugging, understanding what in DETAILS was in the context.

I think the choices are either KEEP or DROP, but a reduced version would not be of any value.

Reply:
Solved in 8a62d0d80d0e4e154a9801a73e3234c99f1bcb0e